spec/spec_helper.rb

```diff
@@ -81,7 +81,80 @@
     # (e.g. via a command-line flag).
     config.default_formatter = 'doc'
   end
+
+  # Check for leaky test resources
+  config.after(:all) do
+    # Exclude acceptable background threads
+    background_threads = Thread.list.reject do |t|
+      group_name = t.group.instance_variable_get(:@group_name) if t.group.instance_variable_defined?(:@group_name)
+      backtrace = t.backtrace || []
+
+      # Current thread
+      t == Thread.current ||
+        # Thread has shut down, but we caught it right as it was still alive
+        !t.alive? ||
+        # Internal JRuby thread
+        defined?(JRuby) && JRuby.reference(t).native_thread.name == 'Finalizer' ||
+        # WEBrick singleton thread for handling timeouts
+        backtrace.find { |b| b.include?('/webrick/utils.rb') } ||
+        # Rails connection reaper
+        backtrace.find { |b| b.include?('lib/active_record/connection_adapters/abstract/connection_pool.rb') } ||
+        # Ruby JetBrains debugger
+        t.class.name.include?('DebugThread') ||
+        # Categorized as a known leaky thread
+        !group_name.nil?
+    end
+
+    unless background_threads.empty?
+      info = background_threads.flat_map do |t|
+        caller = t.instance_variable_get(:@caller) || '(not recorded)'
+        [
+          "#{t} (#{t.class.name})",
+          ' == Caller ==',
+          caller,
+          ' == Backtrace ==',
+          t.backtrace,
+          "\n"
+        ]
+      end.join("\n")
+
+      # We cannot fail tests gracefully in an `after(:all)` block.
+      # The test results have already been decided by RSpec.
+      # We resort to a more "blunt approach.
```

Comment:
Have you considered running the check for background threads and checking it after each individual test?

Reply:
This is what I wanted, but I could not find a way to run code after all `after` RSpec callbacks are run. Creating an `after` callback inserts it at the "end" of the current callback list, but it gets placed before some of the host tests `after` callbacks depending of declaration order.

I debugged the test runs, stepping into RSpec code, and I don't see any places where RSpec would be able to call our code after all `after` blocks are done. We can "inject" (read monkey patch) these changes into RSpec, but there's not supported way for it 😢